src/cfnlint/rules/resources/HardCodedArnProperties.py

```diff
@@ -0,0 +1,34 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+import re
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class HardCodedArnProperties(CloudFormationLintRule):
+    """Checks Resources if ARNs use correctly placed Pseudo Parameters instead of hardcoded Partition, Region, and Account Number"""
+    id = 'I3042'
+    shortdesc = 'ARNs should use correctly placed Pseudo Parameters'
+    description = 'Checks Resources if ARNs use correctly placed Pseudo Parameters instead of hardcoded Partition, Region, and Account Number'
+    source_url = ''
+    tags = ['resources']
+
+    def match(self, cfn):
+        """Check CloudFormation Resources"""
+        matches = []
+        pattern = re.compile(r'arn:(\$\{[^:]*::[^:]*}|[^:]*):[^:]+:(\$\{[^:]*::[^:]*}|[^:]*):(\$\{[^:]*::[^:]*}|[^:]*)')
+        resources = cfn.template.get('Resources', {})
+        if resources:
+            for resourcename, val  in resources.items():
+                candidates = pattern.findall(str(val))

```

Comment:
I think the way this is written should prevent finding other things from triggering this check.  I'm thinking join `!Join`?

Reply:
Agreed !Join wouldn't be caught by this. !Join is a difficult case to handle because there is a lot of variance how joins could be written. Examples I can quickly think of.

```
!Join
  - ''
  - - 'arn:'
    - !Ref AWS::Partition
    - ':s3:::elasticbeanstalk-*-'
    - !Ref 'AWS::AccountId'
``` 
```
!Join [ "", [ 'arn:' ,!Ref 'AWS::Partition',:s3:::bucket]]
```

```
!Join [ ":", [ 'arn' ,!Ref 'AWS::Partition',s3:::bucket]]
```

```
!Join [ ":", [ 'arn' ,!Ref 'AWS::Partition',s3:: , bucket]]
```

Reply:
I should have been more clear.  How it is written is probably ideal.  To your point it becomes tricky.  I was mostly making note for it for myself to know the code wouldn't fail on a join.